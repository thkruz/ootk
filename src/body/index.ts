export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
