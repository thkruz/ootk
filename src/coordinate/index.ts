export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
