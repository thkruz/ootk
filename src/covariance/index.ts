export * from './CovarianceSample.js';
export * from './StateCovariance.js';
