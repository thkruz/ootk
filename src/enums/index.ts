export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
