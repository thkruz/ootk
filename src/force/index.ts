export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

