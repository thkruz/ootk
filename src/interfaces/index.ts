export type { BaseObjectParams } from './BaseObjectParams.js';
export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { SatelliteParams } from './SatelliteParams.js';
export type { SensorParams } from './SensorParams.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { RfSensorParams } from './RfSensorParams.js';
