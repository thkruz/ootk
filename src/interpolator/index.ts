export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

