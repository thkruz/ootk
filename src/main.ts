/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
