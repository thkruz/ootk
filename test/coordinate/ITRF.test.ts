import { Earth, EpochUTC, ITRF, Kilometers, KilometersPerSecond, Vector3D } from './../../src/main';
// Generated by CodiumAI

describe('ITRF', () => {
  let epochUtc: EpochUTC;
  let exampleDate: Date;
  let itrf: ITRF;

  beforeEach(() => {
    const mockExampleDate = new Date(1705109326817);

    exampleDate = new Date(mockExampleDate.getTime());
    epochUtc = EpochUTC.fromDateTime(mockExampleDate);
    itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(1000 as Kilometers, 2000 as Kilometers, 3000 as Kilometers),
      new Vector3D(10 as KilometersPerSecond, 20 as KilometersPerSecond, 30 as KilometersPerSecond),
    );
  });

  // can get the name of the ITRF coordinate system
  it('should return the name of the coordinate system', () => {
    expect(itrf.name).toMatchSnapshot();
  });

  // can determine if the coordinate system is inertial
  it('should return whether the coordinate system is inertial', () => {
    expect(itrf.inertial).toMatchSnapshot();
  });

  /*
   * can get the height of the ITRF coordinate above the surface of the Earth in
   * kilometers
   */
  it('should return the height above the surface of the Earth in kilometers', () => {
    expect(itrf.height).toMatchSnapshot();
  });

  // can get the altitude in kilometers
  it('should return the altitude in kilometers', () => {
    expect(itrf.alt).toMatchSnapshot();
  });

  // can convert the current coordinate to the J2000 coordinate system
  it('should convert the current coordinate to the J2000 coordinate system', () => {
    expect(itrf.toJ2000()).toMatchSnapshot();
  });

  // can convert the current ITRF coordinate to Geodetic coordinate
  it('should convert the current ITRF coordinate to Geodetic coordinate', () => {
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the center of the Earth
  it('should handle position at the center of the Earth', () => {
    itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );
    expect(itrf.toJ2000()).toMatchSnapshot();
  });

  // can handle position at the North Pole
  it('should handle position at the North Pole', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, Earth.radiusPolar),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(itrf.toJ2000()).toMatchSnapshot();
    // Test the toGeodetic method
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the South Pole
  it('should handle position at the South Pole', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, -Earth.radiusPolar as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    itrf.position = new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, -Earth.radiusPolar as Kilometers);
    expect(itrf.toJ2000()).toMatchSnapshot();
    // Test the toGeodetic method
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the International Date Line
  it('should handle position at the International Date Line', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(Earth.radiusEquator, 0 as Kilometers, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    // Test height
    expect(itrf.height).toMatchSnapshot();

    // Test alt
    expect(itrf.alt).toMatchSnapshot();

    // Test toJ2000
    expect(itrf.toJ2000()).toMatchSnapshot();

    // Test toGeodetic
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the Prime Meridian
  it('should handle position at the Prime Meridian', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, Earth.radiusEquator, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(itrf.name).toMatchSnapshot();
    expect(itrf.inertial).toMatchSnapshot();
    expect(itrf.height).toMatchSnapshot();
    expect(itrf.alt).toMatchSnapshot();
    expect(itrf.toJ2000()).toMatchSnapshot();
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at high altitude
  it('should handle position at high altitude', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, (Earth.radiusEquator + 1000) as Kilometers, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(itrf.height).toMatchSnapshot();
    expect(itrf.alt).toMatchSnapshot();
    expect(itrf.toJ2000()).toMatchSnapshot();
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at low altitude
  it('should handle position at low altitude', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, (Earth.radiusEquator - 10) as Kilometers, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(itrf.height).toMatchSnapshot();
    expect(itrf.alt).toMatchSnapshot();
    expect(itrf.toJ2000()).toMatchSnapshot();
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the antimeridian
  it('should handle position at the antimeridian', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D(0 as Kilometers, 0 as Kilometers, Earth.radiusEquator),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    // Test the height property
    expect(itrf.height).toMatchSnapshot();

    // Test the alt property
    expect(itrf.alt).toMatchSnapshot();

    // Test the toJ2000 method
    expect(itrf.toJ2000()).toMatchSnapshot();

    // Test the toGeodetic method
    expect(() => itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at the equator
  it('should handle position at the equator with negative y-coordinate', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(0 as Kilometers, -Earth.radiusEquator as Kilometers, 0 as Kilometers),
      new Vector3D(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(itrf.name).toMatchSnapshot();
    expect(itrf.inertial).toMatchSnapshot();
    expect(itrf.height).toMatchSnapshot();
    expect(itrf.alt).toMatchSnapshot();
    expect(itrf.toJ2000()).toMatchSnapshot();
    expect(itrf.toGeodetic()).toMatchSnapshot();
  });

  // can handle position at different epochs
  it('should handle position at different epochs', () => {
    const position = new Vector3D<Kilometers>(1000 as Kilometers, 2000 as Kilometers, 3000 as Kilometers);
    const velocity = new Vector3D(10 as KilometersPerSecond, 20 as KilometersPerSecond, 30 as KilometersPerSecond);
    const epoch1 = EpochUTC.fromDateTime(exampleDate);
    const epoch2 = EpochUTC.fromDateTime(new Date(exampleDate.getTime() + 1000));
    const itrf1 = new ITRF(epoch1, position, velocity);
    const itrf2 = new ITRF(epoch2, position, velocity);

    // Test height property
    expect(itrf1.height).toMatchSnapshot();
    expect(itrf2.height).toMatchSnapshot();

    // Test alt property
    expect(itrf1.alt).toMatchSnapshot();
    expect(itrf2.alt).toMatchSnapshot();

    // Test toJ2000 method
    const j2000 = itrf1.toJ2000();

    expect(j2000).toMatchSnapshot();

    // Test toGeodetic method
    const geodetic = itrf1.toGeodetic();

    expect(geodetic).toMatchSnapshot();
  });

  // toClassicalElements
  it('should return classical elements', () => {
    const itrf = new ITRF(
      epochUtc,
      new Vector3D<Kilometers>(1000 as Kilometers, 2000 as Kilometers, 3000 as Kilometers),
      new Vector3D(10 as KilometersPerSecond, 20 as KilometersPerSecond, 30 as KilometersPerSecond),
    );

    expect(() => itrf.toClassicalElements()).toThrow();
  });
});
