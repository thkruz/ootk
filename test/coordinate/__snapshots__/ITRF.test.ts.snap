// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ITRF should convert the current ITRF coordinate to Geodetic coordinate 1`] = `
Geodetic {
  "alt": -2622.6838742181076,
  "lat": 0.935735641017013,
  "lon": 1.1071487177940904,
}
`;

exports[`ITRF should convert the current coordinate to the J2000 coordinate system 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -2127.189919581918,
    "y": -667.2073608873604,
    "z": 3004.978765916125,
  },
  "velocity": Vector3D {
    "x": -21.223238500045827,
    "y": -6.827700074091167,
    "z": 30.049679644550032,
  },
}
`;

exports[`ITRF should handle position at different epochs 1`] = `-2622.7069029008003`;

exports[`ITRF should handle position at different epochs 2`] = `-2622.7069029008003`;

exports[`ITRF should handle position at different epochs 3`] = `-2622.7069029008003`;

exports[`ITRF should handle position at different epochs 4`] = `-2622.7069029008003`;

exports[`ITRF should handle position at different epochs 5`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -2127.189919581918,
    "y": -667.2073608873604,
    "z": 3004.978765916125,
  },
  "velocity": Vector3D {
    "x": -21.223238500045827,
    "y": -6.827700074091167,
    "z": 30.049679644550032,
  },
}
`;

exports[`ITRF should handle position at different epochs 6`] = `
Geodetic {
  "alt": -2622.6838742181076,
  "lat": 0.935735641017013,
  "lon": 1.1071487177940904,
}
`;

exports[`ITRF should handle position at high altitude 1`] = `1000`;

exports[`ITRF should handle position at high altitude 2`] = `1000`;

exports[`ITRF should handle position at high altitude 3`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -5313.7806839984505,
    "y": -5118.639766380342,
    "z": 12.531790472796377,
  },
  "velocity": Vector3D {
    "x": 0.37325612599463637,
    "y": -0.38748808327640394,
    "z": -0.0008553895575912965,
  },
}
`;

exports[`ITRF should handle position at high altitude 4`] = `
Geodetic {
  "alt": 1000,
  "lat": 0,
  "lon": 1.5707963267948966,
}
`;

exports[`ITRF should handle position at low altitude 1`] = `-10`;

exports[`ITRF should handle position at low altitude 2`] = `-10`;

exports[`ITRF should handle position at low altitude 3`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -4586.372260974545,
    "y": -4417.944366643129,
    "z": 10.816301918115789,
  },
  "velocity": Vector3D {
    "x": 0.32216074473221884,
    "y": -0.3344444760297924,
    "z": -0.0007382944785579285,
  },
}
`;

exports[`ITRF should handle position at low altitude 4`] = `
Geodetic {
  "alt": -10,
  "lat": 0,
  "lon": 1.5707963267948966,
}
`;

exports[`ITRF should handle position at the International Date Line 1`] = `0`;

exports[`ITRF should handle position at the International Date Line 2`] = `0`;

exports[`ITRF should handle position at the International Date Line 3`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -4424.870330550322,
    "y": 4593.587094016537,
    "z": 10.140457479062466,
  },
  "velocity": Vector3D {
    "x": -0.33496872909712727,
    "y": -0.32266748652491783,
    "z": 0.0007899757588086132,
  },
}
`;

exports[`ITRF should handle position at the International Date Line 4`] = `
Geodetic {
  "alt": 0,
  "lat": 0,
  "lon": 0,
}
`;

exports[`ITRF should handle position at the North Pole 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 14.787499061093714,
    "y": 0.21172360852619188,
    "z": 6356.734413210253,
  },
  "velocity": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`ITRF should handle position at the North Pole 2`] = `
Geodetic {
  "alt": 0,
  "lat": 1.5707963267948966,
  "lon": 0,
}
`;

exports[`ITRF should handle position at the Prime Meridian 1`] = `"ITRF"`;

exports[`ITRF should handle position at the Prime Meridian 2`] = `false`;

exports[`ITRF should handle position at the Prime Meridian 3`] = `0`;

exports[`ITRF should handle position at the Prime Meridian 4`] = `0`;

exports[`ITRF should handle position at the Prime Meridian 5`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -4593.574324568842,
    "y": -4424.881944858347,
    "z": 10.833286953310731,
  },
  "velocity": Vector3D {
    "x": 0.32266663959620306,
    "y": -0.3349696602599569,
    "z": -0.0007394538357761064,
  },
}
`;

exports[`ITRF should handle position at the Prime Meridian 6`] = `
Geodetic {
  "alt": 0,
  "lat": 0,
  "lon": 1.5707963267948966,
}
`;

exports[`ITRF should handle position at the South Pole 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -14.787499061093714,
    "y": -0.21172360852619188,
    "z": -6356.734413210253,
  },
  "velocity": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`ITRF should handle position at the South Pole 2`] = `
Geodetic {
  "alt": 0,
  "lat": -1.5707963267948966,
  "lon": 0,
}
`;

exports[`ITRF should handle position at the antimeridian 1`] = `21.384683407853117`;

exports[`ITRF should handle position at the antimeridian 2`] = `21.384683407853117`;

exports[`ITRF should handle position at the antimeridian 3`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 14.837245536162833,
    "y": 0.21243586576184958,
    "z": 6378.119038744385,
  },
  "velocity": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`ITRF should handle position at the antimeridian 4`] = `[Function]`;

exports[`ITRF should handle position at the center of the Earth 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
  "velocity": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`ITRF should handle position at the equator with negative y-coordinate 1`] = `"ITRF"`;

exports[`ITRF should handle position at the equator with negative y-coordinate 2`] = `false`;

exports[`ITRF should handle position at the equator with negative y-coordinate 3`] = `0`;

exports[`ITRF should handle position at the equator with negative y-coordinate 4`] = `0`;

exports[`ITRF should handle position at the equator with negative y-coordinate 5`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 4593.574324568842,
    "y": 4424.881944858347,
    "z": -10.833286953310731,
  },
  "velocity": Vector3D {
    "x": -0.32266663959620306,
    "y": 0.3349696602599569,
    "z": 0.0007394538357761064,
  },
}
`;

exports[`ITRF should handle position at the equator with negative y-coordinate 6`] = `
Geodetic {
  "alt": 0,
  "lat": 0,
  "lon": -1.5707963267948966,
}
`;

exports[`ITRF should return the altitude in kilometers 1`] = `-2622.7069029008003`;

exports[`ITRF should return the height above the surface of the Earth in kilometers 1`] = `-2622.7069029008003`;

exports[`ITRF should return the name of the coordinate system 1`] = `"ITRF"`;

exports[`ITRF should return whether the coordinate system is inertial 1`] = `false`;
